haystack/document_store/sql.py

```diff
@@ -166,6 +173,8 @@ def get_all_documents(
             DocumentORM.text,
             DocumentORM.vector_id
         ).filter_by(index=index)
+        if page_number is not None and page_size is not None:
+            documents_query = documents_query.offset(page_number * page_size).limit(page_size)
```

Comment:
Pagination via `offset` can cause performance issue as DB have to scan through all the previous rows.
Window query (many major DB have support for Postgres, Oracle, MySQL etc) or Non-Window query using LIMIT (for SQLite) will give good performance advantage. For more information: https://github.com/sqlalchemy/sqlalchemy/wiki/RangeQuery-and-WindowedRangeQuery

Reply:
This would be easy to implement if `get_all_documents()` returns an iterator that yields all documents with a windowed query. As per the current implementation in this PR, each paginated query is an independent call to `get_all_documents()`.  Having said that, it might be a good idea to implement an optional mode where `get_all_documents()` returns an iterator with paginated results. This would abstract away the details of pagination and allow us to use different pagination for each document store while maintaining a uniform interface with `get_all_documents()`.